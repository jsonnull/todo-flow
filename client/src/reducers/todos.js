// @flow

import type { Todos, Todo, Id, Text, TodosAction } from '../types/todos';

const createTodo = (id: Id, text: Text): Todo => ({
  id,
  text,
  completed: false
});

const toggleTodo = (todos: Todos, id: Id): Todos =>
  todos.map(t => (t.id !== id ? t : { ...t, completed: !t.completed }));

const todos = (state: Todos = [], action: TodosAction): Todos => {
  switch (action.type) {
    case 'ADD_TODO':
      return [...state, createTodo(action.id, action.text)];
    case 'TOGGLE_TODO':
      return toggleTodo(state, action.id);
    default:
      return state;
  }
};

export default todos;
