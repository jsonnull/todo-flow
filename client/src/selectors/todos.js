import { createSelector } from 'reselect';

const todosSelector = state => state.todos;
const visibilityFilterSelector = state => state.visibilityFilter;

export const visibleTodosSelector = createSelector(
  todosSelector,
  visibilityFilterSelector,
  (todos, visibilityFilter) => {
    switch (visibilityFilter) {
      case 'SHOW_COMPLETED':
        return todos.filter(t => t.completed);
      case 'SHOW_ACTIVE':
        return todos.filter(t => !t.completed);
      case 'SHOW_ALL':
      default:
        return todos;
    }
  }
);
