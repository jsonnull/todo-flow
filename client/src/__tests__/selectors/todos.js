import { visibleTodosSelector } from '../../selectors/todos';

describe('todos reducer', () => {
  let state;
  beforeEach(() => {
    state = {
      todos: [
        { id: 0, text: 'Test Todo', completed: false },
        { id: 0, text: 'Test Todo', completed: false }
      ],
      visibilityFilter: 'SHOW_ALL'
    };
  });

  it('should handle SHOW_ALL', () => {
    expect(visibleTodosSelector(state)).toMatchSnapshot();
  });

  it('should handle SHOW_ACTIVE', () => {
    state.visibilityFilter = 'SHOW_ACTIVE';
    expect(visibleTodosSelector(state)).toMatchSnapshot();
  });

  it('should handle SHOW_COMPLETED', () => {
    state.visibilityFilter = 'SHOW_COMPLETED';
    expect(visibleTodosSelector(state)).toMatchSnapshot();
  });
});
