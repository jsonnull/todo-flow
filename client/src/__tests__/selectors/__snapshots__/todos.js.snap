exports[`todos reducer should handle SHOW_ACTIVE 1`] = `
Array [
  Object {
    "completed": false,
    "id": 0,
    "text": "Test Todo",
  },
  Object {
    "completed": false,
    "id": 0,
    "text": "Test Todo",
  },
]
`;

exports[`todos reducer should handle SHOW_ALL 1`] = `
Array [
  Object {
    "completed": false,
    "id": 0,
    "text": "Test Todo",
  },
  Object {
    "completed": false,
    "id": 0,
    "text": "Test Todo",
  },
]
`;

exports[`todos reducer should handle SHOW_COMPLETED 1`] = `Array []`;
